mergify_engine/rules.py

```diff
@@ -130,15 +128,25 @@ def get_branch_rule(g_repo, branch, ref=github.GithubObject.NotSet):
     except voluptuous.MultipleInvalid as e:
         raise InvalidRules(".mergify.yml is invalid: %s" % str(e))
 
-    dict_merge(rule, rules.get("default", {}))
+    rule = copy.deepcopy(DEFAULT_RULE)
+    if "default" in rules and rules["default"] is not None:
+        dict_merge(rule, rules["default"])
 
+    found_at_least_one_match = False
     for branch_re in rules.get("branches", []):
         if re.match(branch_re, branch):
             if rules["branches"][branch_re] is None:
                 LOG.info("Rule for %s branch: %s" % (branch, rule))
                 return None
             else:
                 dict_merge(rule, rules["branches"][branch_re])
+                found_at_least_one_match = True
```

Comment:
(nit) you could use a `for/else` I guess by replacing this by a `break.`

Reply:
I just realised that something suck here, imagine say you have:

```
branches:
  ^stable/.*:
      ..
  stable/foobar:
       ...
```

and the branch is `stable/foobar`. Then both rules are merged.

But because `branches` is a dict and python dict are not ordered, the merge order is kind rolling a dice...

Reply:
I also fixed that.